#181414755 Ft user registration

itisWasp opened the pull request:
# #181414755 On successful registration, users should receive a JWT

### What does this PR do?
Register a user and generate a jwt token when an account is successful created.
### Description of Task to be completed?
Once a Users is registered via the registration endpoint, receives a JWT to be required on all subsequent calls to all other endpoints that require authentication.

The User receives a descriptive error message when they enter invalid parameters in an attempt to signup.
if a user has an 
existing user email,
invalid email or password, he is prompted with the appropriate error message. on the other hand the 
password of length less than 8, and a password must contain an Upper case, a Number, a special character to be recognised as strong.

### How should this be manually tested?
Get started By Running commands bellow:
  - clone repo
  - switch to develop branch
  - ``yarn install``: to install all dependencies
  - ``yarn dev``: to start the development server
  -   Open postman
  - Navigate to ``http://localhost:3000/api/v2/users/register``: use ``POST`` http verb  to input the data from the body to be sent to the server example:``{
    "FirstName": "Jane",
    "LastName": "Doe",
    "Email": "test@gmail.com",
    "UserName" : "Doe",
    "Password" : "Password@2022"
}``
  - Create .env file
  - add database url using "DB_CONNECT" as variable
- Navigate to heroku page of the team and open Heroku Postgress DB go to the settings and check the URI of the DB paste it in  "DB_CONNECT" variable.

### What are the relevant pivotal tracker stories?
[#181414755](https://www.pivotaltracker.com/story/show/181414755)

Robertishimwe reviewed (changes requested):
<img width="635" alt="iiiiiiiiiii" src="https://user-images.githubusercontent.com/81079680/158462860-00745736-6bae-4887-a46c-7ed2a0b5a6ea.PNG">

1. firstName & lastName should not accept spaces as value
2. on successfully sign up, user should receive a JWT as the response not a message

Robertishimwe reviewed (commented):
<img width="630" alt="kkkkkkkkkkkkk" src="https://user-images.githubusercontent.com/81079680/158465093-716213a2-9fb3-420f-b64b-02e5e2ec199a.PNG">

**basing on message above, this password should have been accepted**

itisWasp commented:
> <img alt="kkkkkkkkkkkkk" width="630" src="https://user-images.githubusercontent.com/81079680/158465093-716213a2-9fb3-420f-b64b-02e5e2ec199a.PNG">
> 
> **basing on message above, this password should have been accepted**

Fixed the Issue I didn't improve the message the password error message should have displayed that it has a special character missing.

itisWasp commented:
> <img alt="iiiiiiiiiii" width="635" src="https://user-images.githubusercontent.com/81079680/158462860-00745736-6bae-4887-a46c-7ed2a0b5a6ea.PNG">
> 
> 1. firstName & lastName should not accept spaces as value
> 2. on successfully sign up, user should receive a JWT as the response not a message

I also added a new value in the JSOn response for the JWT token.

Robertishimwe reviewed (commented):
this route is note defined
<img width="627" alt="isss" src="https://user-images.githubusercontent.com/81079680/158543680-96081329-d52f-4493-9831-810a2b80892a.PNG">

itisWasp commented:
> this route is note defined <img alt="isss" width="627" src="https://user-images.githubusercontent.com/81079680/158543680-96081329-d52f-4493-9831-810a2b80892a.PNG">

I changed the route but check again with the comment above with the setup of where to check the route.

ericus123 commented on src/controllers/user.js:
```diff
@@ -1,13 +1,48 @@
+        },
+        process.env.TOKEN_SECRET,
+        { expiresIn: process.env.TOKEN_EXPIRE }
+      );
```
This should be reusable.
Create a service for it

itisWasp replied:
> This should be reusable. Create a service for it

Fixed this Issue.

ericus123 commented on src/controllers/user.js:
```diff
@@ -1,13 +1,48 @@
-      res.status(200).send({ Response: data });
+  static createUser = async (req, res) => {
+    //check if a user is in the database
+    const emailExists = await User.findOne({
```
This check should not be in a controller.
Create a helper middleware to check if the user exists.

itisWasp replied:
> This check should not be in a controller. Create a helper middleware to check if the user exists.

Fixed The Issue.

Procech20 commented on src/database/migrations/20220309061357-create-user.js:
```diff
@@ -1,35 +1,39 @@
+'use strict';
```
Please remove this. It's not important as we are using ES6 and causes linting error

itisWasp replied:
> Please remove this. It's not important as we are using ES6 and causes linting error

Fixed This Issue.

ericus123 commented on package.json:
```diff
@@ -1,64 +1,66 @@
-    },
-    "dependencies": {
-        "@babel/cli": "^7.17.6",
-        "@babel/core": "^7.8.4",
-        "@babel/node": "^7.8.4",
-        "@babel/plugin-transform-runtime": "^7.17.0",
-        "@babel/preset-env": "^7.8.4",
-        "@babel/runtime": "^7.17.8",
-        "babel-plugin-istanbul": "^6.0.0",
-        "c8": "^7.11.0",
-        "chai": "^4.3.6",
-        "chai-http": "^4.3.0",
-        "cors": "^2.8.5",
-        "cross-env": "6",
-        "dotenv": "^16.0.0",
-        "express": "^4.17.3",
-        "mocha": "^9.2.2",
-        "nyc": "^15.1.0",
-        "pg": "^8.7.3",
-        "pg-hstore": "^2.3.4",
-        "sequelize": "^6.17.0",
-        "sequelize-cli": "^6.4.1",
-        "supertest": "^6.2.2",
-        "swagger-jsdoc": "^6.1.0",
-        "swagger-ui-express": "^4.3.0"
-    },
-    "devDependencies": {
-        "@babel/eslint-parser": "^7.17.0",
-        "@babel/register": "^7.8.3",
-        "eslint": "8.1.0",
-        "eslint-config-airbnb-base": "^15.0.0",
-        "eslint-config-prettier": "^8.5.0",
-        "eslint-plugin-import": "^2.25.4",
-        "eslint-plugin-prettier": "^4.0.0",
-        "husky": "^7.0.0",
-        "istanbul": "^0.4.5",
-        "nodemon": "^2.0.15",
-        "prettier": "^2.6.0"
-    }
-}
```
Why was this whole file changed?

ericus123 commented on src/controllers/user.js:
```diff
@@ -1,13 +1,61 @@
+    const check = {
+      email: req.body.Email,
+    };
+    // const emailExists = await User.findOne({
+    //   where: { email: req.body.Email },
+    // });
```
Remove this if it's not being used

ericus123 commented on src/services/TokenService.js:
```diff
@@ -0,0 +1,13 @@
+//     },
+//     process.env.TOKEN_SECRET,
+//     { expiresIn: process.env.TOKEN_EXPIRE }
+//   );
```
Remove this

Procech20 commented on src/database/models/user.js:
```diff
@@ -1,25 +1,33 @@
+     * The `models/index` file will call this method automatically.
+     */
+    static associate(models) {
+      // define association here
```
Please remove these comments as they are not necessary!

Procech20 commented on src/swagger/index.js:
```diff
@@ -13,31 +13,97 @@ const paths = {
-docrouter.use('/', serve, setup(specs, { explorer: true }));
+docrouter.use('/', serve, setup(options));
 
 export default docrouter;
```
there's a way you can do the documentation outside this file to avoid too much codes in one file.

itisWasp replied:
Which way do you suggest?

Procech20 commented on src/swagger/index.js:
```diff
@@ -13,31 +13,97 @@ const paths = {
-docrouter.use('/', serve, setup(specs, { explorer: true }));
+docrouter.use('/', serve, setup(options));
 
 export default docrouter;
```
You can do it in an external file and import it in the paths just above the responses.

Commit:
feature(User Registration): successfully register a user

- Document feature using swagger
- Input Validations before saving a user
- Generating a jwt on successfully Registration
- Encrypted Password
- Tested the feature with Mocha and Chai

[Starts #181414755]

Ft User Registration

Rebase from Develop

Update user.js

Fixed Validation Issues

Fixed Issue and Added JWT

Rebased From develop

Rebased from develop

Fixed Error on Linting

Updated The Controller

Rebased from develop

Added New Migration files

Rebased from develop

Added User Registration

Update user.js

Fixed Validation Issues

Fixed Issue and Added JWT

Added User Registration

Fixed Issue and Added JWT

Fixed Error on Linting

Updated The Controller

Fixed Validation

Added Babel Eslint Parser

Rebase from Develop

Rebased From develop

Rebased from develop

Rebased from develop

Added New Migration files

Rebased from develop

Added User Registration

Added User Registration

Fixed Error on Linting

Fixed Validation

Added Babel Eslint Parser

Rebased & Squashed

Rebase from Develop

Update user.js

Fixed Validation Issues

Fixed Issue and Added JWT

Rebased From develop

Rebased from develop

Fixed Error on Linting

Updated The Controller

Rebased from develop

Added New Migration files

Rebased from develop

Added User Registration

Update user.js

Fixed Validation Issues

Fixed Issue and Added JWT

Added User Registration

Fixed Issue and Added JWT

Fixed Error on Linting

Updated The Controller

Fixed Validation

Added Babel Eslint Parser

Rebase from Develop

Rebased From develop

Rebased from develop

Rebased from develop

Added New Migration files

Rebased from develop

Added User Registration

Added User Registration

Fixed Error on Linting

Fixed Validation

Added Babel Eslint Parser

Rebased & Squashed

Added Tests

Fixed test and rebased

Fixed Tests

Fixed Tests 2

Added SSL Fix

Changed Heroku configs

Fixed SSL

Removed migrations

Fixed Tests

Added New email

Added Token_secret

Added Extra Secrets

Fixed Issues

Fixed feedback

Forgot this

Added More Tests

niyojuste commented on package.json:
```diff
@@ -17,7 +17,7 @@
     "lint": "eslint src/** test/** --fix",
     "prepare": "husky install",
-    "test": "cross-env NODE_ENV=test yarn seed && c8 --reporter=text --reporter=lcov ./node_modules/.bin/mocha --timeout 100000 --exit --require @babel/register"
+    "test": "cross-env NODE_ENV=test yarn down && cross-env NODE_ENV=test yarn migrate && cross-env NODE_ENV=test c8 --reporter=text --reporter=lcov ./node_modules/.bin/mocha --timeout 100000 --exit --require @babel/register"
```
Since the use for the dummy test is done, can you find a way to only use beforeEach and afterEach methods

niyojuste commented on src/database/migrations/20220309061357-create-user.js:
```diff
@@ -22,6 +22,12 @@ module.exports = {
+      isVerified: {
+        type: Sequelize.BOOLEAN,
+      },
       createdAt: {
```
New columns should have a migration of their own

ericus123 replied:
@itisWasp work on this

niyojuste commented on test/usertest.spec.js:
```diff
@@ -1,14 +1,168 @@
+        .end((err, res) => {
+          res.should.have.status(400);
+          res.body.should.be.a('object');
+          res.body.should.have.property('message');
```
To be more specific, can you assert the value of this message.

niyojuste commented on test/usertest.spec.js:
```diff
@@ -1,14 +1,168 @@
+        .end((err, res) => {
+          res.should.have.status(400);
+          res.body.should.be.a('object');
+          res.body.should.have.property('message');
```
Same here. You should assert the value of this message

niyojuste commented on test/usertest.spec.js:
```diff
@@ -1,14 +1,168 @@
+        .end((err, res) => {
+          res.should.have.status(400);
+          res.body.should.be.a('object');
+          res.body.should.have.property('message');
```
Same as the previous validation errors, you should assert the value of the message returned

niyojuste commented on test/usertest.spec.js:
```diff
@@ -1,14 +1,168 @@
+        .end((err, res) => {
+          res.should.have.status(400);
+          res.body.should.be.a('object');
+          res.body.should.have.property('message');
```
You should assert the value of this message as well.

niyojuste commented on src/helpers/CheckUser.js:
```diff
@@ -0,0 +1,10 @@
+
+const { User } = models;
+
+const checkUser = async (params) => {
```
Based on the organization of the files, wouldn't this be a UserService as well?

itisWasp replied:
@niyojuste I am using it as a helper function  not the service plz check how it's being used.
